fluster/test.py

```diff
@@ -51,57 +44,172 @@ def __init__(
+        """Compare MD5 hash results."""
+        expected = self.test_vector.result.lower()
+        actual = result.lower()
+
+        if expected == actual:
+            self.test_vector_result.test_result = TestVectorResult.SUCCESS
+        else:
+            self.test_vector_result.test_result = TestVectorResult.FAIL
+
+        self.assertEqual(expected, actual, self.test_vector.name)
+
+
+class PixelComparisonTest(Test):
+    """Test class for pixel comparison"""
+
+    def __init__(
+        self,
+        decoder: Decoder,
+        test_suite: Any,  # can't use TestSuite type because of circular dependency
+        test_vector: TestVector,
+        skip: bool,
+        output_dir: str,
+        reference: bool,
+        timeout: int,
+        keep_files: bool,
+        verbose: bool,
+        reference_decoder: Decoder,
+    ):
+        super().__init__(
+            decoder,
+            test_suite,
+            test_vector,
+            skip,
+            output_dir,
+            reference,
+            timeout,
+            keep_files,
+            verbose,
+        )
+        self._keep_files_during_test = True
+        self.reference_decoder = reference_decoder
+        self.reference_filepath = normalize_path(os.path.join(self.output_dir, self.test_vector.name + "_ref.yuv"))
+
+    def _decode_reference(self) -> str:
+        """Decode the reference file."""
+        keep_files_for_decode = self._keep_files_during_test or self.keep_files
+
+        return self.reference_decoder.decode(
+            self.input_filepath,
+            self.reference_filepath,
+            self.test_vector.output_format,
+            self.timeout,
+            self.verbose,
+            keep_files_for_decode,
+        )
+
+    def _cleanup_reference_file(self) -> None:
+        """Clean up reference file if keep_files is False."""
+        if not self.keep_files and os.path.exists(self.reference_filepath):
+            os.remove(self.reference_filepath)
+
+    def compare_result(self, result: str, output_filepath: str, input_filepath: str) -> None:
+        """Compare decoded output with reference decoder output pixel-wise."""
+        try:
+            reference_result = self._decode_reference()
+
+            if not os.path.exists(self.reference_filepath) and os.path.exists(reference_result):
+                self.reference_filepath = reference_result
+
+            comparison_result = compare_byte_wise_files(
+                self.reference_filepath, output_filepath, keep_files=self.keep_files
+            )
+
+            if comparison_result == 0:
+                self.test_vector_result.test_result = TestVectorResult.SUCCESS
+            else:
+                self.test_vector_result.test_result = TestVectorResult.FAIL
+                error_message = f"Found {comparison_result} pixels that don't meet tolerance"
+                raise ValueError(error_message)
+        finally:
+            self._cleanup_reference_file()
```

Comment:
This cleanup is handled by the base class, you should override _cleanup_if_needed and chain up